arbor/morph/locset.cpp

```diff
@@ -116,6 +117,138 @@ std::ostream& operator<<(std::ostream& o, const terminal_& x) {
     return o << "(terminal)";
 }
 
+// Translate locations in locset distance μm in the proximal direction
+struct proximal_translate_: locset_tag {
+    proximal_translate_(const locset& ls, double distance): start(ls), distance(distance) {}
+    locset start;
+    double distance;
+};
+
+mlocation_list thingify_(const proximal_translate_& dt, const mprovider& p) {
+    const auto& m = p.morphology();
+    const auto& e = p.embedding();
+
+    std::vector<mlocation> L;
+
+    const auto start = thingify(dt.start, p);
+    const auto distance = dt.distance;
+
+    for (auto loc: start) {
+        auto distance_remaining = distance;
+
+        while (loc.branch != mnpos) {
+            const auto branch = loc.branch;
+            const auto branch_length = e.branch_length(branch);
+            const auto dist_pos = loc.pos;
+            const auto prox_pos = dist_pos - distance_remaining / branch_length;
+
+            if (prox_pos>=0) {
+                // The target is inside this branch.
+                L.push_back({branch, prox_pos});
+                break;
+            }
+            if (auto parent = m.branch_parent(branch); parent==mnpos) {
+                // The root has been reached; return the start of the branch attached to root.
+                L.push_back({branch, 0.});
+                break;
+            }
+            else {
+                loc = {parent, 1.};
+            }
+            distance_remaining -= dist_pos*branch_length;
+        }
+    }
+
+    return L;
+}
+
+locset proximal_translate(locset ls, double distance) {
+    return locset(proximal_translate_{ls, distance});
+}
+
+std::ostream& operator<<(std::ostream& o, const proximal_translate_& l) {
+    return o << "(proximal-translate " << l.start << " " << l.distance << ")";
+}
+
+// Translate locations in locset distance μm in the distal direction
+struct distal_translate_: locset_tag {
+    distal_translate_(const locset& ls, double distance): start(ls), distance(distance) {}
+    locset start;
+    double distance;
+};
+
+locset distal_translate(locset ls, double distance) {
+    return locset(distal_translate_{ls, distance});
+}
+
+mlocation_list thingify_(const distal_translate_& dt, const mprovider& p) {
+    const auto& m = p.morphology();
+    const auto& e = p.embedding();
+
+    std::vector<mlocation> L;
+
+    auto start = thingify(dt.start, p);
+    auto distance = dt.distance;
+
+    struct branch_interval {
+        msize_t bid;
+        double distance;
+    };
+
+    for (auto c: start) {
+        std::stack<branch_interval> branches_reached;
```

Comment:
Does this need to be a stack? It doesn't hurt, but we just sort the generated locsets in the end.

Reply:
The algorithm effectively uses a depth-first search to traverse the sub-trees, for which a stack is the natural fit.
The actual branch sections in the output are stored in a different data structure: the `std::vector` `L`.